examples/load/lift.linux-aarch64.json

```diff
@@ -0,0 +1,56 @@
+{
+  "scie": {
+    "lift": {
+      "name": "cowsay",
+      "base": "~/.nce",
+      "files": [
+        {
+          "name": "get.sh"
+        },
+        {
+          "name": "cowsay-1.1.0.jar",
+          "key": "cowsay.jar",
+          "size": 1724250,
+          "hash": "212ee64546eb8b5074572fed4107d850eb90bc462aa3099c0ac8ea63fdca7811",
+          "type": "blob",
+          "source": "get"
+        },
+        {
+          "name": "openjdk-19.0.1_linux-aarch64_bin.tar.gz",
+          "key": "jdk",
+          "size": 194660832,
+          "hash": "88cadc91d5c7c540ea9df5d23678bb65dc2092fe4e00650b39d87f24f2328e17",
+          "type": "tar.gz",
+          "source": "get"
+        }
+      ],
+      "boot": {
+        "commands": {
+          "": {
+            "exe": "{jdk}/jdk-19.0.1/bin/java",
+            "args": [
+              "-jar",
+              "{cowsay.jar}"
+            ],
+            "env": {
+              "=JAVA_HOME": "{jdk}/jdk-19.0.1",
+              "=PATH": "{jdk}/jdk-19.0.1/bin:{scie.env.PATH}"
+            }
+          }
+        },
+        "bindings": {
+          "get": {
+            "exe": "{get.sh}",
+            "args": [
+              "{scie.lift}"
```

Comment:
I'm not sure I follow the 1st paragraph. The `{scie.lift}` placeholder is a placeholder that is expanded to the absolute path of the lift manifest itself, after having been extracted from the scie file itself. In other words, the presence of a `{scie.lift}` placeholder triggers that extraction. It happens to be done in `~/.nce/<lift manifest hash>/lift.json`.

My confusion over the 1st paragraph likely makes the 2nd paragraph track less well than it should so I'll stop there to hear what you're thinking in response.

A thing to keep in mind though is that the ultimate idea here is definitely not to use `bash` + `jq` + `curl` to fetch file contents - that depends way too much on the host OS to work at all See the variance in the windows example for a highlight of this fragility. Instead the `a-scie/ptex` repo will provide a static binary for all platforms supported by `a-scie/jump` the does all this. No `bash`, `jq` or `curl` required.

Reply:
Sorry, let me be clearer. 

As I understand it, the value of the `name` field is implicitly tacked on to the end of these `args`. I was proposing that this be explicit, e.g., using `"{scie.file.name}"`.

And then in paragraph 2 I expanded that to the realization we could have `"{scie.file.XXX}"` be a placeholder for any field in the file's object. And if we then allow custom fields in there, you could, for example, stick a `url` field in there, and reference it in `args` as `"{scie.file.url}"`. 

I got that `{scie.lift}` is a placeholder to the manifest itself, but right now we pass in the entire manifest, so you must use json parsing (whether `jq`, or in ptex implementation, or in whatever other mechanism) to pull out the single URL you care about, based on the name.  Which is cumbersome, and also means that the binding binary must know about the shape of the manifest. What I'm suggesting is that we could instead do the more obvious thing, and just pass the URL (and maybe other standard or custom fields, as needed) right in via `args`.

Of course one could still use `{scie.lift}` in the above manner if one wished, if the binary did need the entire manifest for some reason.

Is that clearer? I'll post an example next.

Reply:
```
{
  "scie": {
    "lift": {
      "name": "cowsay",
      "base": "~/.nce",
      "files": [
        {
          "name": "get.sh"
        },
        {
          "name": "cowsay-1.1.0.jar",
          "key": "cowsay.jar",
          "size": 1724250,
          "hash": "212ee64546eb8b5074572fed4107d850eb90bc462aa3099c0ac8ea63fdca7811",
          "type": "blob",
          "url": "https://repo1.maven.org/maven2/com/github/ricksbrown/cowsay/1.1.0/cowsay-1.1.0.jar",
          "source": "get"
        },
        {
          "name": "openjdk-19.0.1_linux-x64_bin.tar.gz",
          "key": "jdk",
          "size": 195925792,
          "hash": "7a466882c7adfa369319fe4adeb197ee5d7f79e75d641e9ef94abee1fc22b1fa",
          "type": "tar.gz",
          "url": "https://download.java.net/java/GA/jdk19.0.1/afdd2e245b014143b62ccb916125e3ce/10/GPL/openjdk-19.0.1_linux-x64_bin.tar.gz",
          "source": "get"
        }
      ],
      "boot": {
        "commands": {
          "": {
            "exe": "{jdk}/jdk-19.0.1/bin/java",
            "args": [
              "-jar",
              "{cowsay.jar}"
            ],
            "env": {
              "=JAVA_HOME": "{jdk}/jdk-19.0.1",
              "=PATH": "{jdk}/jdk-19.0.1/bin:{scie.env.PATH}"
            }
          }
        },
        "bindings": {
          "get": {
            "exe": "{get.sh}",
            "args": [
              "{scie.file.url}"
            ]
          }
        }
      }
    }
  }
}
```

And  `get.sh` is:

```
#!/usr/bin/env bash
set -euo pipefail

url="$1"
exec curl -fL "${url}"
```

Reply:
And of course there is nothing magical about the `url` key - it's a custom field that scie ignores, other than to look it up for placeholder replacement. Its name could be anything.

Reply:
Right, gotcha. The thing is, right now, the "scie" portion of the manifest is designed strict. No unknown fields are allowed except as top-level lift manifest object keys. I'd need to poke holes in that or else treat "url" as a first class file field, sort-of negating the abstraction.

Reply:
I assume we both agree even the slightly trimmed down `get.sh` your idea allows for is still system-dependent. The goal here is total independence with the statically compiled ptex binary doing all the work.

Reply:
Maybe I should ask what issue you're worried about with the current setup that allowing custom file fields and reference of them makes better. The downside of custom file fields and placeholders supporting them is pretty much just complicating JSON parsing code and complicating placeholder.rs placeholder parsing code. The advantage of the PR approach is there is no new complication in either arena, all the existing abstractions are enough.

Reply:
Thinking don in the weeds a bit, I think the trickiest bit of using placeholders in this way is they become, for the 1st time, indeterminate. It's not `{scie.file[0].url}`, but `{scie.file.url}` for the contextually appropriate file.

Reply:
The issue was mostly the added complexity and constraints on the manifest and the binary that provides the binding. If we expect that to ~always be `ptex` then I guess that's fine. But we'd probably want a simple `curl` example to work (and we can bundle curl in the scie, presumably, to make that self-contained)

What I'm proposing seems like a more straightforward interface to me, and that's what you asked for feedback on... It just seems weird to me to say "we arbitrarily don't allow custom keys in this part of the manifest, but we do in this other part, so any binding needs to be able to parse the manifest".

I think there are two paths here:

1) This is intended to be general-purpose, ptex is just an example of its use. In that case, let's support that ergonomically?
2) This is really just for ptex. In that case, we can bake in that support without requiring the binding boilerplate which will in practice be identical in every usage?

Right now, it seems like half-way between these two, from a usability perspective.

Reply:
The intention is definitely general purpose. I thought the fact the existing `{scie.lift}` placeholder made this all work was a good proof of that. The speed bump of parsing your own custom metadata did not raise itself as a burden in my eyes, but you're pointing out it is. I do agree it would be nicer to streamline that away if possible. It does get tricky though when you get down to details. If a file accepts arbitrary extra keys, your example is only streamlined for the case when the key values are scalars. In this example, "url" -> str. So either the JSON handling of that would need to enforce scalars only (str, number or bool with a convention around bool - the JSON convention), or else if the value is an array or object, revert to passing that value as a serialized json blob I guess that the consumer must then parse.

I'll think on this some more.

Reply:
I realize I may not share intuitions about boilerplate. Perhaps to clear that up, but perhaps not, I was viewing lift manifests as something no human would write. That would be done by other tools. It's intended that https://github.com/a-scie/lift would be the actual CLI interface humans would use to assemble scies if not leaning on Pants or Pex to do the dirty work. So I have not been considering too much about boilerplate reduction, etc in the lift manifest as of yet. As such, JSON was an easy choice since serde handles it so well and its resulting parser / emitter is so small. Editing these by hand is not fun, I continuously mess up trailing commas.

So, from my perspective, your boilerplate concern is a good one, but only in regards to the burdens placed on the binding command implementations; i.e.: the fact they currently almost without fail will need to have code to parse json to get at extra metadata. Its that, in my mind, that would be good to make easier.

Reply:
OK, that is good context. I was definitely imagining humans writing manifests.

Reply:
With "humans will not commonly write these" in mind, I think it's OK as-is. I just want to make sure the "fetching a URL" case is relatively straightforward, and it sounds like ptex will handle that. So the need to use a custom binding is pretty rare, and at that point we can say "you have to parse JSON".

Reply:
Ok, thanks for thinking through this with me. I did try a bit here this morning to implement the arbitrary field attachments to File and its not trivial at least. Although adding support for passing parameters in this way will be backward compatible, the actual call API of a `source=<binding command>` will not, since the file name argument that's currently passed will either need to disappear or remain redundant and ignored. I'm going to proceed with the existing structure of all custom metadata is accessed through top level keys and see how reality plays out. Hopefully I can feel confident about all this prior to 1.0.